### PR TITLE
Allow to use find next and find previous commands during say all.

### LINKED ISSUE #11563: Say all is interrupted when using find next/previous commands
### Steps to reproduce:

* In keyboard options, enable "Allow skim reading" option
* Go to any webpage, e.g.: https://www.nvaccess.org/
* Press NVDA+control+F to search a term, e.g. "NVDA".
*Press NVDA+DownArrow to start SayAll
* Press NVDA+F3

### Actual behavior:

NVDA:
* sets the virtual cursor on the next occurrence of the searched word
* reads a little bit of text (from the searched word to the end of the line)
* stops reading

### Expected behavior:

NVDA:
* sets the virtual cursor on the next occurrence of the searched word
* continues reading in say all from the new position of the virtual cursor onward, i.e. from the next occurrence of the searched word onward.

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

2020.2

#### Windows version:

Win 10

#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Not now but I remember it has always behaved this way.
#### If addons are disabled, is your problem still occuring?
Yes
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
No

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:

Fixes #11563

### Summary of the issue:

In browse mode, the find next/previous commands stop text reading if executed during say all. Since these commands move the caret and read the line at the new position, this would make sense that they do not interrupt say all.

### Description of how this pull request fixes the issue:

* specified "resumeSayAllMode" for findNext and findPrevious scripts.
* adapt doFindText with an extra argument specifiying if sayAll should resume.
* also converted the 2 modified scripts to the new way to write scripts with script decorator.

### Testing performed:

* Tested that findNext and findPrevious do not interrupt sayAll if skim reading option is enabled.
* Tested that findNext and findPrevious work as before if skim reading option is disabled.
* Test that help mode on these 2 scripts still announce their function.

### Known issues with pull request:

None

### Change log entry:

Section: Changes

`When reading with say all in browse mode, the find next and find previous commands do not stop reading anymore if Allow skim reading option is enabled; say all rather resumes from after the next or previous found term. (#11563)`

### CODE REVIEW

**Comment by lukaszgo1** [comment]
@CyrilleB79 Please convert modified scripts to the scriptDecorator fully i.e. specify gesture assigned to them inside decorator and not in a __gestures__ dict.

**Comment by CyrilleB79** [comment]
Thanks @lukaszgo1 for this review and feedback. The decorator syntax should now be fully converted.
I still need to address the failing AppVeyor check. I do not know why this fail; I should investigate when I have more time.

File: source/cursorManager.py [hunk 1]
```diff
@@ -174,21 +175,41 @@ def run():
...
-
+		self.doFindText(
+			self._lastFindText,
+			caseSensitive=self._lastCaseSensitivity,
+			willSayAllResume=willSayAllResume(gesture),
+		)
+
+	@script(
+		description=_(
+			"find the previous occurrence of the previously entered text string from the current cursor's position"
```
Comment (lukaszgo1) [inline source/cursorManager.py:198]: This message is missing translator commend hence the build failure. For a discussion about making such problems more obvious see #10928

**Comment by CyrilleB79** [comment]
@lukaszgo1 thanks for review. I had found the cause of the issue on my side and did not see your review suggestion in parallel. Anyway thanks also for the pointer on the issue regarding failing checkpot test.

**Comment by feerrenrut** [comment]
Generally this looks ok to me, but what isn't clear on the issue, or in the changes string, is where the say all should resume from. Should users expect say-all to resume speaking the content prior to the interruption, or continue speaking text after the found term?

**Comment by CyrilleB79** [comment]
@feerrenrut thanks for review.

You wrote:
> Generally this looks ok to me, but what isn't clear on the issue, or in the changes string, is where the say all should resume from. Should users expect say-all to resume speaking the content prior to the interruption, or continue speaking text after the found term?

Say all continues reading after the found term. This seemed obvious to me... but it seems it was not the case for all.
I have updated the corresponding issue (#11563) with more details in actual result and expected result.

I will also modify the change log item. Would the change log item be clearer as follow?

`When reading with say all in browse mode, the find next and find previous commands do not interrupt reading anymore if Allow skim reading option is enabled; say all rather resumes from the next or previous found term. (#11563)

It's a bit long, and maybe my English formulation is not so good. Any more concise proposal is welcome.
When the formulation is confirmed, I will update the initial description of this PR with the new proposal for change log.

**Comment by feerrenrut** [comment]
Thanks for confirming, this is what I expected it to do also, but you never know! The change log entry looks good to me. Better slightly verbose than not clear.

I don't seem to be able to get this to work with a laptop keyboard layout. It worked as expected with desktop layout.

**Comment by feerrenrut** [comment]
Ah, it does work with laptop layout, but I used the wrong say-all command:

Name | Desktop key | Laptop key | Touch | Description
-- | -- | -- | -- | --
Say all with review | numpadPlus | NVDA+shift+a | 3-finger flick down (text mode) | Reads from the current position of the review cursor, moving it as it goes

I don't think this should hold back this PR, but it would be good to have at least an issue open to document it. Would you be interested in investigating whether making this work during say all with review makes sense?

**Review by feerrenrut — APPROVED** [review]
Thanks @CyrilleB79

**Comment by CyrilleB79** [comment]
> Ah, it does work with laptop layout, but I used the wrong say-all command:
> Name 	Desktop key 	Laptop key 	Touch 	Description
> Say all with review 	numpadPlus 	NVDA+shift+a 	3-finger flick down (text mode) 	Reads from the current position of the review cursor, moving it as it goes
> 
> I don't think this should hold back this PR, but it would be good to have at least an issue open to document it. Would you be interested in investigating whether making this work during say all with review makes sense?

No this presently do not make sense. Indeed, the Find command applies to virtual cursor (as quick jump key), not to review cursor. If one day there is a search function using the review cursor in NVDA, this behavior may be replicated in the case of say all with review cursor.
I do not know however if such a feature would be very useful. But I am not the best person to judge of it since I usually work review cursor tethered to system cursor.

@lukaszgo1, I think you usually work with both cursors separately. Could you comment further if required and open an issue describing the need if you feel it applicable?

**Comment by lukaszgo1** [comment]
> No this presently do not make sense. Indeed, the Find command applies to virtual cursor (as quick jump key), not to review cursor. If one day there is a search function using the review cursor in NVDA, this behavior may be replicated in the case of say all with review cursor.
> I do not know however if such a feature would be very useful. But I am not the best person to judge of it since I usually work review cursor tethered to system cursor.
> 
> @lukaszgo1, I think you usually work with both cursors separately. Could you comment further if required and open an issue describing the need if you feel it applicable?



Having find command working with the review cursor might be useful in some cases and it is something which I intent to implement one day if time permits (see #7855). When this is implemented we can reconsider, but I personally don't use say all with review at all so it is unlikely I'll be interested in pursuing this further.

**Pull request merged**